src/general_utils.erl

```diff
@@ -0,0 +1,74 @@
+%%%-------------------------------------------------------------------
+%%% @author kevin
+%%% @copyright (C) 2020, <COMPANY>
+%%% @doc
+%%%
+%%% @end
+%%% Created : 28. Mär 2020 10:56
+%%%-------------------------------------------------------------------
+-module(general_utils).
+-author("kevin").
+-include("gingko.hrl").
+
+%% API
+-export([group_by/2, add_to_value_list_or_create_single_value_list/3, sorted_insert/3, get_or_default_dict/3, get_or_default_map_list/3, get_or_default_map_list_check/3, concat_and_make_atom/1, atom_replace/3]).
+
+%% @doc Takes function that groups entries form the given list in a dictionary
+%%      For example grouping a list of journal entries by txid to get all journal entries that belong to a certain txid
+-spec group_by(fun((ListType :: term()) -> GroupKeyType :: term()), [ListType :: term()]) -> dict:dict(GroupKeyType :: term(), ListType :: term()).
+group_by(Fun, List) ->
+  lists:foldr(
+    fun({Key, Value}, Dict) ->
+      dict:append(Key, Value, Dict)
+    end, dict:new(), [{Fun(X), X} || X <- List]).
+
+%% @doc Updates a dictionary that stores lists as values
+%%      If a list exists for the given key then the given value is added to the list otherwise a new list with the given value is added to the dictionary
+-spec add_to_value_list_or_create_single_value_list(DictionaryTypeAToValueTypeBList :: dict:dict(KeyTypeA :: term(), TypeBList :: [term()]), KeyTypeA :: term(), ValueTypeB :: term()) -> dict:dict(KeyTypeA :: term(), TypeBList :: [term()]).
+add_to_value_list_or_create_single_value_list(DictionaryTypeAToValueTypeBList, KeyTypeA, ValueTypeB) ->
+  dict:update(KeyTypeA, fun(ValueTypeBList) ->
+    [ValueTypeB | ValueTypeBList] end, [ValueTypeB], DictionaryTypeAToValueTypeBList).
+
+-spec sorted_insert(TypeA :: term(), [TypeA :: term()], fun((TypeA :: term(), TypeA :: term()) -> boolean())) -> [TypeA :: term()].
```

Comment:
You can use (to some limited extend) polymorphic types in Erlang. This might support readability here.

Reply:
Check Sec 7.5 [here](https://erlang.org/doc/reference_manual/typespec.html)